CI Mutants: Action Membership Run

ASuciuX opened the pull request:
This is the PR enabling the full manual dispatch of the mutants CI so that they could run for more than 6 hours. It has a custom action limiting the access to trigger that only to the wanted team.

This should be merged only after:
1. the Master PR is merged which enables a GUI button to trigger this and make it testable https://github.com/stacks-network/stacks-core/pull/5082
2. the composite action is merged and fully functional https://github.com/stacks-network/actions/pull/50

wileyj commented:
waiting on successful run from changes in #5082 before merging this. changes appear fine

ASuciuX commented:
- added attribution membership action & format output in this PR https://github.com/stacks-network/actions/pull/51, but isn't a blocker for the current one
- it passed default checks
https://github.com/stacks-network/stacks-core/actions/runs/10423875818
- it fails as expected if the user is not part of the team ( I updated it to check for devops team and I am not part of it ) https://github.com/stacks-network/stacks-core/actions/runs/10424084011

wileyj reviewed (dismissed):
shipit

obycode commented on .github/workflows/pr-differences-mutants.yml:
```diff
@@ -9,16 +9,43 @@ on:
+      ignore_timeout:
+        description: "Ignore mutants timeout limit"
+        required: false
+        type: choice
+        options:
+          - true
```
Instead of a choice with just one option, I think you can use `type: boolean`, or if you stick with `choice` I think you need to add a false option.

wileyj replied:
good catch - i don't recall seeing boolean as an option in the past, but sure enough it is one:
https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/accessing-contextual-information-about-workflow-runs#inputs-context

ASuciuX replied:
If I enable the false option (by adding it or changing the type to boolean), developers outside the @stacks-network/blockchain-team who have write access to create PRs would be able to trigger CI flows lasting less than 6 hours using the `false` input. While I don't expect this to have a major impact, since they can already initiate CI mutants workflows through their PRs, I wanted to double check with you if this behavior is intended.

obycode commented on .github/workflows/pr-differences-mutants.yml:
```diff
@@ -9,16 +9,43 @@ on:
+          GITHUB_TOKEN: ${{ secrets.GH_TOKEN }}
+
+      - name: Fail if the user does not have the right permissions
+        if: ${{ inputs.ignore_timeout == 'true' && steps.check_right_permissions.outputs.is_team_member != 'true' }}
```
If you make the change above to make this a boolean, then I think you need this corresponding change.
```suggestion
        if: ${{ inputs.ignore_timeout && steps.check_right_permissions.outputs.is_team_member != 'true' }}
```

wileyj replied:
i see. in that case then - why not set an env var after you check if the user is part of a team, and use that env var (our output var, either would work) in later conditionals?

that way, you could remove the workflow_dispatch choice option entirely.

ASuciuX replied:
that's a nice idea, but aren't the automatic workflow runs limited to 6 hours? the main goal would be to run for more than 6 hours and if that requires a manual trigger i think workflow_dispatch would be the easiest way to do this.

----
if the automatic workflow runs aren't limited to 6 hours, I would do these 2 actions:
1. give unlimited time for prs done by blockchain-team developers
2. keep the dispatch in case the pr is done by someone outside the blockchain-team so that someone from the team could trigger it manually, if that is the case

ASuciuX replied:
or are you referring specifically just to remove the true/false input on dispatch and let it take the context from who calls the workflow_dispatch? (true - part of blockchain-team, false - otherwise )

wileyj replied:
yes - if the person running this via workflow_dispatch is on the allowed team - we can bypass the time limit of 6 hours that we've set. 
else, we keep that time limit, and we can remove the choice option from displayed (we won't need it - either the calling actor is in the team or they aren't, and the time limit can set accordingly). 

so:
- workflow_dispatch called manually by someone *not on* allowed team: timeout at 6 hours
- workflow_dispatch called manually by someone *on* allowed team: no timeout set
- workflow triggered automatically (i.e. as part of a PR): timeout at 6 hours

ASuciuX replied:
i removed it, but there is an unknown. how does it has the context to differentiate between manual workflow_dispatch and automatic pr? (`2` and `3`)

previously i gave it this context through the input field that i added. on PRs it would be false and on workflow_dispatch true. now it proceeds on both cases the same way, running it as it sees that the action is done by a member of the allowed team, not knowing it should stop directly on the automatic PR trigger.

- this workflow dispatch work as expected https://github.com/stacks-network/stacks-core/actions/runs/10488274882
- this automatic PR trigger should have stopped directly and display that a manual trigger is required as it will timeout after 6 hours https://github.com/stacks-network/stacks-core/actions/runs/10488267327

wileyj replied:
> i removed it, but there is an unknown. how does it has the context to differentiate between manual workflow_dispatch and automatic pr? (`2` and `3`)
> 
> previously i gave it this context through the input field that i added. on PRs it would be false and on workflow_dispatch true. now it proceeds on both cases the same way, running it as it sees that the action is done by a member of the allowed team, not knowing it should stop directly on the automatic PR trigger.
> 
> * this workflow dispatch work as expected https://github.com/stacks-network/stacks-core/actions/runs/10488274882
> * this automatic PR trigger should have stopped directly and display that a manual trigger is required as it will timeout after 6 hours https://github.com/stacks-network/stacks-core/actions/runs/10488267327

to diff, you would need to modify the steps that run conditionally (or when you set an env var etc) by checking if the workflow was triggered manually (workflow_dispatch). ex:
```
github.event_name == 'workflow_dispatch'
```
https://github.com/stacks-network/stacks-core/blob/ci/custom-action-membership/.github/workflows/ci.yml#L116

once you know what triggered the workflow, you can adjust the environment as you need it. 

I *think* that's what you're asking at least. if not, ping me

ASuciuX replied:
awesome, i've missed that. i will update it

ASuciuX replied:
Worked as expected:
- stops on PR trigger on member if more mutants than threshold  https://github.com/stacks-network/stacks-core/actions/runs/10491998635
- running on workflow_dispatch CI https://github.com/stacks-network/stacks-core/actions/runs/10492055238

wileyj commented:
changes after addressing comments from @obycode and the conditionals i recommended look good. 
However, i think these changes should also be merged into master else they won't take effect until the next time a release branch is merged there (it may be a few weeks at the earliest). 

could you cherry-pick these commits in a new PR based off of master to override #5082 ?


otherwise, these changes look good to me

ASuciuX commented:
> could you cherry-pick these commits in a new PR based off of master to override #5082 ?  

sure, https://github.com/stacks-network/stacks-core/pull/5100

obycode reviewed (approved):
Makes sense 👍

blockstack-devops commented:
This pull request has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
